File: projeto - Cópia/front/src/socketService.js
import { io } from 'socket.io-client';
import API_BASE from './api';

let socket = null;

// Função para inicializar a conexão do Socket.IO
export const initSocket = (token) => {
  if (socket) {
    socket.disconnect();
  }

  socket = io(API_BASE, {
    query: { token },
    path: '/socket.io',
    transports: ['websocket', 'polling'],
    autoConnect: true
  });

  socket.on('connect', () => {
    console.log('Socket.IO conectado com sucesso');
  });

  socket.on('connect_error', (error) => {
    console.error('Erro de conexão Socket.IO:', error.message);
  });

  socket.on('disconnect', (reason) => {
    console.log(`Socket.IO desconectado: ${reason}`);
  });

  return socket;
};

// Obter a instância do Socket.IO
export const getSocket = () => {
  return socket;
};

// Entrar em um tópico
export const joinTopic = (topicoId) => {
  if (socket && socket.connected) {
    socket.emit('joinTopic', topicoId);
  }
};

// Sair de um tópico
export const leaveTopic = (topicoId) => {
  if (socket && socket.connected) {
    socket.emit('leaveTopic', topicoId);
  }
};

// Desconectar o Socket.IO
export const disconnectSocket = () => {
  if (socket) {
    socket.disconnect();
    socket = null;
  }
};

export default {
  initSocket,
  getSocket,
  joinTopic,
  leaveTopic,
  disconnectSocket
};

File: projeto_mobile/lib/services/socket_service.dart
import 'package:socket_io_client/socket_io_client.dart' as IO;
import 'package:shared_preferences/shared_preferences.dart';

class SocketService {
  static final SocketService _instance = SocketService._internal();
  factory SocketService() => _instance;
  SocketService._internal();

  IO.Socket? _socket;
  bool _isConnected = false;

  bool get isConnected => _isConnected;

  Future<void> connect() async {
    if (_socket != null && _isConnected) return;

    try {
      final prefs = await SharedPreferences.getInstance();
      final token = prefs.getString('token');

      _socket = IO.io('https://your-api-url.com', <String, dynamic>{
        'transports': ['websocket'],
        'autoConnect': false,
        'query': {'token': token},
      });

      _socket!.connect();

      _socket!.onConnect((_) {
        print('Socket conectado');
        _isConnected = true;
      });

      _socket!.onDisconnect((_) {
        print('Socket desconectado');
        _isConnected = false;
      });

      _socket!.onConnectError((error) {
        print('Erro na conexão socket: $error');
        _isConnected = false;
      });

    } catch (e) {
      print('Erro ao inicializar socket: $e');
    }
  }

  void disconnect() {
    if (_socket != null) {
      _socket!.disconnect();
      _socket!.dispose();
      _socket = null;
      _isConnected = false;
    }
  }

  // Métodos para tópicos
  void joinTopic(int topicoId) {
    if (_socket != null && _isConnected) {
      _socket!.emit('joinTopic', topicoId);
    }
  }

  void leaveTopic(int topicoId) {
    if (_socket != null && _isConnected) {
      _socket!.emit('leaveTopic', topicoId);
    }
  }

  // Métodos para temas
  void joinTema(int temaId) {
    if (_socket != null && _isConnected) {
      _socket!.emit('joinTema', temaId);
    }
  }

  void leaveTema(int temaId) {
    if (_socket != null && _isConnected) {
      _socket!.emit('leaveTema', temaId);
    }
  }

  // Método para ouvir eventos
  void on(String event, Function(dynamic) callback) {
    if (_socket != null) {
      _socket!.on(event, callback);
    }
  }

  // Método para remover listeners
  void off(String event) {
    if (_socket != null) {
      _socket!.off(event);
    }
  }

  // Método para emitir eventos
  void emit(String event, dynamic data) {
    if (_socket != null && _isConnected) {
      _socket!.emit(event, data);
    }
  }
}